From: Kefeng Wang <wangkefeng.wang@huawei.com>
Subject: mm: compaction: convert to folio_isolate_movable()
Date: Mon, 26 Aug 2024 12:01:30 +0800

The tail page will always fail to isolate for non-lru-movable and LRU page
in isolate_migratepages_block(), so move PageTail check ahead, then
convert to use folio_isolate_movable() helper and more folios.

Link: https://lkml.kernel.org/r/20240826040132.1202297-3-wangkefeng.wang@huawei.com
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Baolin Wang <baolin.wang@linux.alibaba.com>
Cc: David Hildenbrand <david@redhat.com>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Zi Yan <ziy@nvidia.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/compaction.c |   32 +++++++++++++++++---------------
 1 file changed, 17 insertions(+), 15 deletions(-)

--- a/mm/compaction.c~mm-compaction-convert-to-folio_isolate_movable
+++ a/mm/compaction.c
@@ -1074,39 +1074,41 @@ isolate_migratepages_block(struct compac
 			}
 		}
 
+		if (PageTail(page))
+			goto isolate_fail;
+
+		folio = page_folio(page);
+
 		/*
 		 * Check may be lockless but that's ok as we recheck later.
-		 * It's possible to migrate LRU and non-lru movable pages.
-		 * Skip any other type of page
+		 * It's possible to migrate LRU and non-lru movable folioss.
+		 * Skip any other type of folios.
 		 */
-		if (!PageLRU(page)) {
+		if (!folio_test_lru(folio)) {
 			/*
-			 * __PageMovable can return false positive so we need
-			 * to verify it under page_lock.
+			 * __folio_test_movable can return false positive so
+			 * we need to verify it under page_lock.
 			 */
-			if (unlikely(__PageMovable(page)) &&
-					!PageIsolated(page)) {
+			if (unlikely(__folio_test_movable(folio)) &&
+					!folio_test_isolated(folio)) {
 				if (locked) {
 					unlock_page_lruvec_irqrestore(locked, flags);
 					locked = NULL;
 				}
 
-				if (isolate_movable_page(page, mode)) {
-					folio = page_folio(page);
+				if (folio_isolate_movable(folio, mode))
 					goto isolate_success;
-				}
 			}
 
 			goto isolate_fail;
 		}
 
 		/*
-		 * Be careful not to clear PageLRU until after we're
-		 * sure the page is not being freed elsewhere -- the
-		 * page release code relies on it.
+		 * Be careful not to clear lru flag until after we're
+		 * sure the folio is not being freed elsewhere -- the
+		 * folio release code relies on it.
 		 */
-		folio = folio_get_nontail_page(page);
-		if (unlikely(!folio))
+		if (unlikely(folio_try_get(folio)))
 			goto isolate_fail;
 
 		/*
_
